src/Microsoft.AspNet.Mvc.Core/Infrastructure/MvcRouteHandler.cs

```diff
@@ -128,31 +134,23 @@ private Task InvokeActionAsync(RouteContext context, ActionDescriptor actionDesc
 
         private void EnsureServices(HttpContext context)
         {
-            if (_actionContextAccessor == null)
+            if (_servicesRetrieved)
             {
-                _actionContextAccessor = context.RequestServices.GetRequiredService<IActionContextAccessor>();
+                return;
             }
 
-            if (_actionInvokerFactory == null)
-            {
-                _actionInvokerFactory = context.RequestServices.GetRequiredService<IActionInvokerFactory>();
-            }
+            // The IActionContextAccessor is optional. We want to avoid the overhead of using CallContext
+            // if possible.
```

Comment:
Is the perf issue w/ `CallContext` on read or write (or both)? If only reading is a major problem, could register the service unconditionally and just not read it anywhere in our code.